File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/MenuHandler.java
package closer.vlllage.com.closer.handler.helpers;

import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.pool.PoolMember;

public class MenuHandler extends PoolMember {
    public void show(MenuOption... menuOptions) {
        $(AlertHandler.class).make()
                .setLayoutResId(R.layout.menu_modal)
                .setOnAfterViewCreated((alertConfig, view) -> {
                    RecyclerView menuRecyclerView = view.findViewById(R.id.menuRecyclerView);
                    menuRecyclerView.setLayoutManager(new LinearLayoutManager($(ActivityHandler.class).getActivity(), LinearLayoutManager.VERTICAL, false));
                    List<MenuOption> options = new ArrayList<>();

                    for (MenuOption option : menuOptions) {
                        if (option.visible) options.add(option);
                    }

                    menuRecyclerView.setAdapter(new MenuOptionAdapter(options, menuOption -> {
                        menuOption.callback.run();
                        alertConfig.getDialog().dismiss();
                    }));
                })
                .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.close))
                .show();
    }

    public static class MenuOption {

        @DrawableRes int iconRes;
        @StringRes int titleRes;
        @NonNull Runnable callback;
        boolean visible = true;

        public MenuOption(@DrawableRes int iconRes, @StringRes int titleRes, @NonNull Runnable callback) {
            this.iconRes = iconRes;
            this.titleRes = titleRes;
            this.callback = callback;
        }

        public MenuOption visible(boolean visible) {
            this.visible = visible;
            return this;
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/MenuHandler.kt
package closer.vlllage.com.closer.handler.helpers

import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.extensions.visible
import com.queatz.on.On
import kotlinx.android.synthetic.main.menu_modal.view.*

class MenuHandler constructor(private val on: On) {
    fun show(vararg menuOptions: MenuOption, title: String? = null, button: String? = null, buttonCallback: (() -> Unit)? = null) {
        on<AlertHandler>().make().apply {
            layoutResId = R.layout.menu_modal
            onAfterViewCreated = { alertConfig, view ->
                val menuRecyclerView = view.menuRecyclerView
                menuRecyclerView.layoutManager = LinearLayoutManager(on<ActivityHandler>().activity, RecyclerView.VERTICAL, false)
                val options = mutableListOf<MenuOption>()

                for (option in menuOptions) {
                    if (option.visible) options.add(option)
                }

                menuRecyclerView.adapter = MenuOptionAdapter(options) { menuOption ->
                    menuOption.callback.invoke()
                    alertConfig.dialog?.dismiss()
                }

                view.title.visible = title != null
                view.title.text = title
            }
            positiveButtonCallback = { buttonCallback?.invoke() }
            positiveButton = button ?: on<ResourcesHandler>().resources.getString(R.string.close)
            show()
        }
    }

    class MenuOption(@DrawableRes internal var iconRes: Int,
                     @StringRes internal var titleRes: Int? = null,
                     internal var title: String? = null,
                     internal var callback: () -> Unit) {
        internal var visible = true

        fun visible(visible: Boolean): MenuOption {
            this.visible = visible
            return this
        }
    }
}
